File: julia/rational-numbers/rational-numbers.jl
struct RationalNumber{Int} <: Real
    num::Int
    denom::Int

    function RationalNumber(n::Int, d::Int=1)
        n == d == 0 && throw(ArgumentError(""))
        return new{Int}(div.(flipsign.((n, d), d), gcd(n,d))...)
    end
end

numerator(a::RationalNumber) = a.num
denominator(a::RationalNumber) = a.denom

Base.Float64(a::RationalNumber) = a.num / a.denom
Base.promote_type(::Type{RationalNumber{Int}}, ::Type{<:Real}) = Float64
Base.promote_type(::Type{<:Real}, ::Type{RationalNumber{Int}}) = Float64

Base.inv(a::RationalNumber) = RationalNumber(a.denom, a.num)
Base.zero(::Type{RationalNumber{Int}}) = RationalNumber(0, 1)
Base.one(::Type{RationalNumber{Int}}) = RationalNumber(1, 1)
Base.abs(a::RationalNumber) = RationalNumber(abs(a.num), a.denom)

Base.:+(a::RationalNumber, b::RationalNumber) = RationalNumber(a.num * b.denom + b.num * a.denom, a.denom * b.denom)
Base.:*(a::RationalNumber, b::RationalNumber) = RationalNumber(a.num * b.num, a.denom * b.denom)
Base.:*(a::RationalNumber, b::Int) = a * RationalNumber(b)
Base.:-(a::RationalNumber, b::RationalNumber) = a + b * -1
Base.:/(a::RationalNumber, b::RationalNumber) = a * inv(b)
Base.:^(a::RationalNumber, p::Int) = RationalNumber(a.num^p, a.denom^p)
Base.:^(a::Int, p::RationalNumber) = (a^Float64(p.num))^inv(Float64(p.denom))

Base.:>(a::RationalNumber, b::RationalNumber) = Float64(a) > Float64(b)
Base.:<(a::RationalNumber, b::RationalNumber) = Float64(a) < Float64(b)
Base.:<=(a::RationalNumber, b::RationalNumber) = a < b || a == b
Base.:>=(a::RationalNumber, b::RationalNumber) = a > b || a == b

Base.show(io::IO, a::RationalNumber) = print(io, "$(a.num)//$(a.denom)")

File: csharp/rational-numbers/RationalNumbers.cs
using System;

using static Helpers;

public struct RationalNumber
{
    public int Num { get; set; }
    public int Denom { get; set; }

    public RationalNumber(int numerator, int denominator = 1)
    {
        var gcd = GCD(numerator, denominator);

        Num = FilpSign(numerator, denominator) / gcd;
        Denom = FilpSign(denominator, denominator) / gcd;
    }

    public static RationalNumber operator +(RationalNumber r1, RationalNumber r2) => new(r1.Num * r2.Denom + r2.Num * r1.Denom, r1.Denom * r2.Denom);

    public static RationalNumber operator -(RationalNumber r1, RationalNumber r2) => new(r1.Num * r2.Denom - r2.Num * r1.Denom, r1.Denom * r2.Denom);

    public static RationalNumber operator *(RationalNumber r1, RationalNumber r2) => new(r1.Num * r2.Num, r1.Denom * r2.Denom);

    public static RationalNumber operator /(RationalNumber r1, RationalNumber r2) => r1 * r2.Inverse();

    public RationalNumber Inverse() => new(Denom, Num);

    public RationalNumber Abs() => new(Math.Abs(Num), Denom);

    public RationalNumber Reduce() => this;

    public RationalNumber Exprational(int power) => new((int)Math.Pow(Num, power), (int)Math.Pow(Denom, power));

    public double Expreal(int baseNumber) => baseNumber.Expreal(this);
}

public static class RealNumberExtension
{
    public static double Expreal(this int realNumber, RationalNumber r) => NthRoot(Math.Pow(realNumber, r.Num), r.Denom);
}

public static class Helpers
{
    public static int GCD(int a, int b)
    {
        (a, b) = (Math.Abs(a), Math.Abs(b));

        while (a != 0 && b != 0)
        {
            if (a > b)
                a %= b;
            else
                b %= a;
        }

        return a | b;
    }

    public static double NthRoot(double a, int n) => Math.Pow(a, 1.0 / n);

    public static int FilpSign(int num, int sign) => num * Math.Sign(sign == 0 ? 1 : sign);
}